brite2caom2/storage_name.py

```diff
@@ -0,0 +1,181 @@
+        return self._file_name.endswith('.ndatdb')
+
+    @property
+    def has_undecorrelated_metadata(self):
+        return self._file_name.endswith('.orig')
+
+    @property
+    def has_data(self):
+        return (
+            self._file_name.endswith('.orig')
+            or self._file_name.endswith('.ndatdb')
+            or self._file_name.endswith('.avedb')
+        )
+
+    @property
+    def is_last_to_ingest(self):
+        return self._file_name.endswith('.rlogdb')
+
+    @property
+    def prev(self):
+        result = f'{self._obs_id}_prev.jpg'
+        if self.has_undecorrelated_metadata:
+            result = f'{self._obs_id}_un_prev.jpg'
+        return result
+
+    @property
+    def thumb(self):
+        result = f'{self._obs_id}_prev_256.jpg'
+        if self.has_undecorrelated_metadata:
+            result = f'{self._obs_id}_un_prev_256.jpg'
+        return result
+
+    def is_valid(self):
+        return True
+
+    def rename(self, original_extension, new_extension):
+        """
+        Common code to refer to a different file than the one currently being processed.
+        :return:
+        """
+        new_fqn = self._source_names[0].replace(original_extension, new_extension)
+        new_uri = self._get_uri(basename(new_fqn))
+        return new_fqn, new_uri
+
+    def set_file_id(self):
+        self._file_id = BriteName.remove_extensions(self._file_name)
+
+    def set_product_id(self):
+        self._product_id = 'decorrelated'
+        if self.has_undecorrelated_metadata:
+            self._product_id = 'un-decorrelated'
+
+    @staticmethod
+    def remove_extensions(f_name):
+        return splitext(f_name)[0]
+
+    @staticmethod
+    def archived(file_name):
+        # files whose purpose is other than storage/ingestion at CADC
+        return not (file_name.endswith('.lst') or file_name.endswith('.md5'))
+
+
+def add_entry(sn, original_extension, new_extension, clients, metadata_reader):
+    """
+    Common code to retrieve and reference a different file than the one currently being processed in the collection
+    of 5 that make up a BRITE-Constellation Observation.
+
+    This happens because the unit of work for any pipeline is a file, but the BRITE-Constellation plane- and
+    artifact-level metadata obtain most of their metadata and data from the .orig file, with a few exceptions:
+      - the TemporalWCS metadata for the decorrelated plane
+      - the previews for the decorrelated plane require data from two files
+    """
+    fqn, uri = sn.rename(original_extension, new_extension)
+    if dirname(fqn) is None or dirname(fqn) == '':
+        fqn = f'./{fqn}'
+    # retrieve the file if it doesn't already exist
+    if not exists(fqn) and clients is not None:
+        clients.data_client.get(dirname(fqn), uri)
+    # retrieve the file metadata if it doesn't already exist
+    with open(fqn) as f:
+        metadata_reader._read_file(f, uri)
+    return uri, fqn
```

Comment:
Name is misleading (maybe `get_entry` instead of `add_entry`)? It seams a bit complicated. Doesn't `BriteName` store the file root (not sure if it's the same with jobs_id)? Can the calling code do `'{sn.file_id}.{new_extension}'` to get the new name of the file?

Reply:
Changed the name to `get_entry`.

Most of the functionality in this method is in place for the case where any of the files associated with an observation, that are all already at CADC, are re-ingested.

I try really hard to NOT do things like `{sn.file_id}.{new_extension}` in random code, because when a telescope changes file locations and file names behaviour, and relationships between files, it ends up being brittle. Note also that this is one of the locations that depends on the method `rename`/`use_different_file`, with its temporal cohesion.